zerver/actions/message_flags.py

```diff
@@ -252,16 +253,21 @@ def do_update_message_flags(
     create_historical_user_messages(user_id=user_profile.id, message_ids=historical_message_ids)
 
     if operation == "add":
-        count = msgs.update(flags=F("flags").bitor(flagattr))
+        msgs = msgs.extra(where=[UserMessage.where_flag_is_absent(flag_int)])
+        updated_message_ids = [um.message_id for um in msgs]
+        msgs.update(flags=F("flags").bitor(flagattr))
     elif operation == "remove":
-        count = msgs.update(flags=F("flags").bitand(~flagattr))
+        msgs = msgs.extra(where=[UserMessage.where_flag_is_present(flag_int)])
+        updated_message_ids = [um.message_id for um in msgs]
+        msgs.update(flags=F("flags").bitand(~flagattr))
```

Comment:
```py
    def where_flag_is_present(flag_int: int) -> str:
        return f"flags & {1 << flag_int} <> 0"
```

Do we need this `flag_int` and left-shift logic? There's already `flagattr`. From the way the existing code uses that, it looks like `where_flag_is_present` can be simplified to `f"flags & {flagattr} <> 0"` -- or to the equivalent in the Django query builder, involving `F("flags").bitand(flagattr)`.

That would also be good because from these definitions:
```py
    flagattr = getattr(UserMessage.flags, flag)
    flag_int = UserMessage.ALL_FLAGS.index(flag)
```
it's not totally clear how these two relate to each other.

Reply:
`flagattr` is actually, `<Bit: number=1, is_set=True>`, I thought it was just a normal bit value. I changed `flag_int` to `flagattr.number`.

Reply:
OK, that's helpful.

I think a further cleanup would be to pass `flagattr` itself into those methods, rather than `flagattr.number` -- if the method wants `.number`, it can get that itself, as an implementation detail. That way the code outside that model is passing around nice semantically-meaningful `Bit` objects instead of integers.

Reply:
Ok, updated thanks!